src/compiler/types.ts

```diff
@@ -2748,6 +2756,8 @@ namespace ts {
     export interface AnonymousType extends ObjectType {
         target?: AnonymousType;  // Instantiation target
         mapper?: TypeMapper;     // Instantiation mapper
+        elementType?: Type;      // Element expressions of evolving array type

```

Comment:
It's separate from `autoArrayType`, whose type is actually TypeReference. 

I'm talking about the comment at the top of the new code that says 

> Evolving array types are ultimately converted into manifest array types
> and never escape the getFlowTypeOfReference function."

I was suggesting something like:

```ts
export interface AutoArrayType extends AnonymousType {
  elementType?: Type;
  finalArrayType?: Type;
}
```

And then having the new code that returns AnonymousType today return AutoArrayType instead.